NETMSG_SECRET_CHAT: Secret chat messages

saurtron opened the pull request:
### Work done

- New `NETMSG_SECRET_CHAT` packet.
  - New type of chat message with some (light) "privacy guarantees".
- New `PLAYER_SECRET_CHAT = 15` autohost message type.
- `InterplayerSecretsForbidden`: server config option, disallows the below modoption. default: true
- `InterplayerSecrets`: modOption to control secrets among players. default: false
- Can be sent through the `/SByNum <id> <msg>` command.
- Also `Spring.SendSecretChat(msg, playerID)`

#### Related issues

- https://github.com/beyond-all-reason/RecoilEngine/issues/2040
- https://github.com/beyond-all-reason/Beyond-All-Reason/issues/5173

### Specs:

- Doesn't get logged into replays.
- By default Players can only send to 255.
  - unless AllowInterplayerSecrets is set, then player to player is allowed.
- Not for groupsend for now, although could be implemented.
- Spectators can't send or receive them.
- Server will forward to the destination player only.
- Appears with "Secret:" prefix in chat.
- They all get forwarded to the Autohost interface.

### How to test

- Set AllowInterplayerSecrets=0 or don't set it
   - Open multiplayer game
   - Do /SByNum 1 hello
      - shouldn't reach the other player
   - Do /SByNum 255 hello
      - reaches the server but you need autohost or some print to see it
   - quit
- Set AllowInterplayerSecrets=1
   - Do /SByNum 1 hello
      - reaches the other player
- When watching replays the secret messages shouldn't be there, also server didn't record it in replay.
- Check with demo parser to see secret messages are not there. 

### Remarks

- Unsure about the specs but the feature is more or less done, probably will need to polish the rules.
- It does go into infolog.txt, might be good to avoid this too.
- Might be good to have a lua api to send them too.
- Maybe should have a shortcode akin to `/w` for sending to usernames, but since more for sending to server not sure it's needed, could also just detect names and ids inside the `<id>` parameter of `/SByNum`.
- Maybe should log into the players own replay if they're the destination or source.

saurtron committed:
MSG_PRIVATE_CHAT

saurtron committed:
More handling of private flag.

saurtron committed:
Don't add to traffic if private.

saurtron committed:
Revert "Don't add to traffic if private."

This reverts commit 4310a5ad2e48d011797fc585b50d43187a1e745b.

saurtron committed:
PRIVATE_CHAT -> SECURE_CHAT.

saurtron committed:
Don't send SECURE messages to everyone.

saurtron committed:
SECURE_CHAT -> SECRET_CHAT.

saurtron committed:
SendDirect -> SendSecret.

saurtron committed:
SendDirect -> SendSecret.

saurtron committed:
Show as 'Secret:'.

saurtron committed:
Server option 'AllowInterplayerSecrets' to control whether players can send secret messages among themselves.

saurtron committed:
Don't allow specs sending secret messages.

saurtron committed:
Also don't allow sending to spectators.

saurtron committed:
Fix compilation.

p2004a commented:
Autohost needs to know if the message is secret or not, at the moment that bit is not forwarded.

saurtron committed:
Change description of the right command, and SaySecureByPlayerIDActionExecutor -> SaySecretByPlayerIDActionExecutor.

saurtron commented:
> Autohost needs to know if the message is secret or not, at the moment that bit is not forwarded.

yeah, was afraid to break something, can I just add that as a 4th parameter inside the asio connection without risk?

saurtron committed:
New autohost message type PLAYER_SECRET_CHAT = 15 for secret messages.

saurtron committed:
Describe the NETMSG_SECRET_CHAT packet type.

saurtron committed:
More documentation for the secret message types.

saurtron commented on rts/Net/Protocol/NetMessageTypes.h:
```diff
@@ -94,6 +94,7 @@ enum NETMSG {
 	NETMSG_GAME_FRAME_PROGRESS= 77, // int32_t frameNum # this special packet skips queue & cache entirely, indicates current game progress for clients fast-forwarding to current point the game #
 
 	NETMSG_PING = 78, // uint8_t playerNum, uint8_t pingTag, float localTime
+	NETMSG_SECRET_CHAT     = 79,  // uint8_t from, dest; std::string message; (same as NETMSG_CHAT, but private)
```
done, yeah that's why I changed terminology to "secret", Improved it now.

saurtron committed:
Fix typo.

sprunk commented on rts/Net/GameServer.cpp:
```diff
@@ -560,6 +562,16 @@ void CGameServer::Broadcast(std::shared_ptr<const netcode::RawPacket> packet)
+{
+	const int fromPlayer = msg.fromPlayer;
+	const int destPlayer = msg.destination;
+
+	if (allowInterplayerSecrets && destPlayer >= 0 && destPlayer < players.size() && !players[fromPlayer].IsSpectator() && !players[fromPlayer].IsSpectator()) {
+		players[destPlayer].SendData(std::shared_ptr<const RawPacket>(msg.Pack()));
+	}
```
What if server (= player 255) does `/sbynum 1`, will this condition filter it or not? If sending `any player -> 255` works then sending `255 -> any player` probably should too (so that the server can send a reply).

saurtron replied:
I'll look into it, server does have a special way to do that so not really an issue (it's the server after all, it can send a direct message any moment), but maybe those can be merged somehow, or at least allow it just in case, sure.

That's unless 255 messages are coming from elsewhere, but I thought they actually generate in the server.

sprunk replied:
If server skips this check then perhaps some sort of `// messages to/from server (255) handled separately by server itself` would be nice here, but otherwise looks good enough.

saurtron replied:
> If server skips this check then perhaps some sort of `// messages to/from server (255) handled separately by server itself` would be nice here, but otherwise looks good enough.

Added the message, but still changed it to allow, so we can send from autohost at some point, I checked and doesn't seem like autohost can signal messages as secret atm, since the protocol has no type flag when it sends back, its just sending messages. Would need to signal it with either `/x` prefix, or a special byte as first element of the packed, by using invisible characters or maybe utf8 forbidden chars, I think it would be feasible.

saurtron committed:
Check the right player.

saurtron committed:
destPlayer -> toPlayer.

saurtron committed:
Actually forward msg secret status to autohost.

saurtron committed:
Improve conditions at CGameServer::SendSecret and document a bit.

saurtron committed:
Lua Spring.SendSecretChat to send secret chats.

saurtron committed:
Make AllowInterplayerSecrets a modoption instead.

- this way all players know about the config automatically and can be configured per game.

saurtron committed:
Check gamesetup to see if sending to player is allowed.

saurtron committed:
do not forward interplayer secrets to autohost if they're not allowed

saurtron committed:
Check if sending to other players is allowed on game command.

saurtron committed:
No use for std::format here.

sprunk commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -628,16 +630,47 @@ int LuaUnsyncedCtrl::SendSpectatorChat(lua_State* L) {
+		return luaL_error(L, std::format("Incorrect arguments to Spring.%s(message string, playerID integer)", __func__).c_str());
 
 	const int playerID = luaL_checkint(L, 2);
 	if (!playerHandler.IsValidPlayer(playerID))
-		return luaL_error(L, "Error in function '%s': Invalid Player ID %d", __func__, playerID);
+		return luaL_error(L, "Error in function '%s': Invalid Player ID %d", __func__);
```
Refactoring mistake? playerID should stay

saurtron committed:
Readd missing playerID.

saurtron commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -628,16 +630,47 @@ int LuaUnsyncedCtrl::SendSpectatorChat(lua_State* L) {
 int LuaUnsyncedCtrl::SendPrivateChat(lua_State* L) {
 	if (lua_gettop(L) != 2 || !lua_isstring(L, 1))
-		return luaL_error(L, "Incorrect arguments to Spring.SendPrivateChat(message string, playerID integer)");
+		return luaL_error(L, std::format("Incorrect arguments to Spring.%s(message string, playerID integer)", __func__).c_str());
```
forgot to upload one commit fixing that, done

saurtron committed:
Restore original form of logically unchanged line.

saurtron committed:
Also fix luaL_error's at SendSecretChat.

saurtron committed:
Fix whitespace.

saurtron committed:
Bring back config variable so interplayer secrets can be completely disabled by the server.

sprunk commented on rts/Game/UnsyncedGameCommands.cpp:
```diff
@@ -566,6 +566,43 @@ class SayPrivateByPlayerIDActionExecutor : public IUnsyncedActionExecutor {
+		if (!gameSetup->interplayerSecrets && playerID != SERVER_PLAYER) {
+			LOG_L(L_WARNING, "Sending secrets to other players disallowed by game configuration");
+		}
+		else if (playerID >= 0) {
```
Can this be some sort of `IsPlayerValid`?

saurtron committed:
Better validation of playerID.

saurtron committed:
Remove uneeded include.

saurtron commented:
Changed the behaviour for how to enable this:

- made interplayerSecrets a modconfig, so it can be set for the game and its status known to all participants, with just a config variable it wouldn't be possible for clients to know without taking extra steps.
- still have a config option for the server to completely disable them and override the modconfig: InterplayerSecretsForbidden.

sprunk commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -634,7 +635,36 @@ int LuaUnsyncedCtrl::SendPrivateChat(lua_State* L) {
+	if (playerID != SERVER_PLAYER && !playerHandler.IsValidPlayer(playerID))
+		return luaL_error(L, "Error in function '%s': Invalid Player ID %d", __func__, playerID);
+
+	if (playerID != SERVER_PLAYER && !gameSetup->interplayerSecrets)
+		return luaL_error(L, "Error in function '%s': Game configuration disallows sending secrets to players", __func__);
+
```
If this is going to error out, it would be good to expose the bool e.g. in LuaConstGame so that one could check for it:
```lua
if Game.interplayerSecrets then
  Spring.SendSecretChat("hello there", greviousID)
else
  Spring.Echo("secrets not allowed but at least there's no error")
end
```

saurtron replied:
can't they check the modoption directly?

sprunk replied:
Hm I somewhat dislike it being a modoption but let's roll with it (there are many precedents anyway). In that case it should probably be listed in `cont/base/springcontent/EngineOptions.lua`.

sprunk replied:
The issue with checking modoption is that if you pass `aLLoWsEcRetS=1` in startscript then engine will read it but Lua will not. And there are other "engine modoptions" so either you get inconsistent behaviour where secrets are case-sensitive but others aren't, or the others get a breaking change where I wouldn't be confident to say it's fine.

The more conceptual issue is that for less experienced devs it's hard to tell what is a gameside modoption and what is handled by engine. This is e.g. why for unitdefs, customparams go in their own subtable where it's 100% separated and engine never reads customparams. It's not that painful so far given that most of the existing ones are rare or obvious enough (maxspeed, fixedallies...) but I think that the new secrets modoption also fits that category so it's not too bad either.

sprunk replied:
Let's stick with the modoption for now and maybe think about doing something to all the engine modoptions later.

saurtron committed:
Describe the InterplayerSecrets modoption at EngineOptions.lua.

p2004a reviewed (commented):
Based on code like `fromPlayer == SERVER_PLAYER` I assume you also wanted to allow server to be able to send messages: that's not currently possible with this PR.

The handling of messages from autohost is here: https://github.com/beyond-all-reason/RecoilEngine/blob/cdfc9d7b872c3d890fc7c77bcb645a23cd9ec8e5/rts/Net/GameServer.cpp#L848-L860

and you can see that autohost can trigger only commands defined in `CGameServer::PushAction`.

---

And to step back a bit, make sure we aren't introducing something legacy nobody really asked for, is private *message* really what we want?

There is also another mechanism how information is exchanged: LuaMsg. Those are not chat, those are structured pieces of information that are supposed to be handled by game side code etc.

So the question is: Wouldn't those be better suited for the original purpose of referenced issues? Where it is not human originating message, but some structured information from game?

p2004a commented:
> But secrets want to be able to send a response to specific playerID. So if we reuse mode for this, it would change meaning depending on script type (mode 97 would mean an individual response to player with ID 97 if script is secret, but "send to allies" if script is Lua).

I was thinking mode adding `mode` value `'p'`, for "private". That means that it's a private LuaMsg send only to that player from the server, and in the opposite direction, from player to the server.

Autohost can't send `LuaMsg` as the moment, it needs a command in `CGameServer::PushAction`, we could add new one, e.g. `/privluamsg <player_name> <data>` (and to follow current ways, also `/privluamsgbynum <player_num> <data>`). Alternatively allow injecting any other LuaUI msgs: `/luauimsg <mode> <player_id/0> <data>`.

saurtron commented:
Converting to draft since getting remade into a completely new packet type or smth in that line. Please voice your implementation ideas.

I'm not sure yet, but what I noticed so far is both luamsg and chat are pretty borked implementations. Luamsg doesn't have a section parameter, and forces everyone to parse content to know who the destination is. Chat also doesn't even have a mechanism to forward or represent destination properly, and I think it even gets LOG and then parsed and sent to players, not even sure how that works yet, anyways the interface is really limited if we can even call it that.

Looks like historic implementations that sticked, so maybe good also to rethink all that. Do we need a packet for "secret player<->server" info, or do we need proper chat and/or luamsg packets and implementations. Not ready to answer but I think are the questions to make at this point.
